### ISSUE TITLE
Reconfigure EM to run via Docker

### ISSUE THREAD

**Issue by darozak** [Darwin-River/Ex-Machinis#14]
We need to reconfigure EM so that it runs within the Docker environment.  This will help us make improvements to the system on one or more development servers before efficiently porting it to the production server.  

Key objectives for this requirement are:

- Set up a functional development server.
- Reconfigure the current code to work within a Docker environment.
- Ensure that the mail server avoids triggering spam filters.
- Create an easily accessible config file that will allow the Docker image to be installed on a different server with a unique domain name.
- Set up a persistent directory that holds key DB and config files so that new updates don't overwrite user and game data.
- Provide detailed documentation on how to manage the config file, update Docker images, and install them on other servers.

**Comment by darozak** [comment]
Set up an Umbutu dev server on Digital Ocean

**Comment by darozak** [comment]
We've abandoned the transition to Docker for now since it turned out to be more complicated than we thought and is not worth the extra effort.

**Comment by darozak** [comment]
Hi Henderb.  Sorry for the delay in getting back to you on this.  I appreciate your follow up on the Docker issue and offer to help us implement a solution.  My main objective in using Docker is to facilitate the instillation of the system on ad hoc dev servers so that multiple coders can work to improve the game.  It would need to be a really simple install solution that would allow other Upwork hires to quickly set up their own Ex Machinis dev server for coding and trouble shooting.  Can we do this with Docker?  It seems like there is a steep learning curve for Docker and that it is more geared towards enterprise solutions.

 

Dave

**Comment by henderbj** [comment]
Docker is some complex because you need to configure all the system. For your ex-machinis, in Docker you need to configure at least:
* Laravel.
* Mysql (as I remind for your DB).
* Ex-machinis code.
* Networking.

The Docker part is easy, only some days (probably couple weeks) learning it. But that is only the tip of the iceberg.

Docker community recommends having simple containers (one service by container), but then the setup can be difficult (you will need to enable and secure network connections on mysql, for example). But, for your development of ex-machinis, you could have a single container with the whole ex-machinis installation.
Another best practice is to use a Dockerfile for the container, but you can also just run a base image (like ubuntu:xenial) and use the container like a machine, by installing and configure all the software inside it, and saving the images for later usage. This is explained in step 2 of getting started page:
https://docs.docker.com/get-started/part2/
When i was working on this, I got stuck last while making Laravel to use our smtp server, and still I don't have that knowledge.
For your rapid development I suggest you make someone who has experience with manual installation of whole ex-machinis system to learn Docker. I could help in the Docker part as open source contributor on my spare time.

In short:
* Laravel is the most important skill needed for this. Docker last.
* Its much more easy to learn Docker than Laravel.
* You can create single container with whole ex-machinis installation.
* You can create the container without Dockerfile, by saving images for later usage.

**Comment by henderbj** [comment]
Hi, guys.

Dave, if you still need to have a fast and congruent development environment, but easy to configure and use, I suggest you to use Laravel  Homestead: 
https://laravel.com/docs/5.8/homestead

Laravel Homestead uses Vagrant, which can work with several backends, being Virtualbox more common.

I specially recommend to use Per-Project Installation (https://laravel.com/docs/5.8/homestead#per-project-installation), instead of one global Homestead.

You probably have used Virtualbox and Vagrant before. If so, then use Laravel Homestead will be really easy to have. If you don't have that experience, however, I bet you will find Vagrant+Virtualbox combo much easier to learn than Docker.

After you have configured Homestead, for one new guy to have Ex-machines running, will be as simple as install Virtualbox, Vagrant, clone your ex-machines repository, and run vagrant up.

If you need any help using it, please let me know.

Kind regards,
Henderb
